src/KubernetesClient/Controllers/ResourceEventDeltaBlock.cs

```diff
@@ -0,0 +1,299 @@
+                QueueActionLocked(resourceEvent);
+            }
+
+            if (_queue.Count == 1) // we've just added to empty queue, kick off processing
+                OfferMessagesToLinks();
+            return DataflowMessageStatus.Accepted;
+        }
+
+        private TKey KeyOf(ResourceEvent<TResource> obj) => KeyOf(obj.Value);
+
+        private TKey KeyOf(TResource obj) => _keyFunc(obj);
+
+
+        /// <summary>
+        ///     Checks if block is marked for completion and marks itself as completed after queue is drained
+        /// </summary>
+        private void SetCompletedIfNeeded()
+        {
+            lock (_lock)
+            {
+                if (!_isCompleting || _queue.Count != 0) return;
+                foreach (var link in _targets.Where(x => x.LinkOptions.PropagateCompletion))
+                {
+                    _log.LogTrace("Completing link");
+                    link.Target.Complete();
+                }
+            }
+
+            _taskCompletionSource.TrySetResult(null);
+        }
+
+
+        private void CombineDeltas(List<ResourceEvent<TResource>> deltas)
+        {
+            if (deltas.Count < 2) return;
+            if (deltas.First().EventFlags.HasFlag(EventTypeFlags.Sync)) // if we had a sync item queued up and got something else, get rid of sync
+                deltas.RemoveAt(0);
+            if (deltas.Count < 2) return;
+
+            // if the entire object was created and removed before worker got a chance to touch it and worker has not chose to see these
+            // types of events, we can just get rid of this "transient" object and not even notify worker of its existence
+            if (_skipTransient && deltas[0].EventFlags.HasFlag(EventTypeFlags.Add) && deltas[deltas.Count - 1].EventFlags.HasFlag(EventTypeFlags.Delete))
+                deltas.Clear();
+        }
+
+        private void QueueActionLocked(ResourceEvent<TResource> obj)
+        {
+            var id = KeyOf(obj);
+            var exists = _items.TryGetValue(id, out var deltas);
+            if (!exists)
+            {
+                deltas = new List<ResourceEvent<TResource>>();
+                _items[id] = deltas;
+                _queue.Enqueue(deltas);
+                _log.LogTrace("Item not already know, adding to queue");
+            }
+
+            deltas.Add(obj);
+            CombineDeltas(deltas);
+        }
+
+        private void OfferMessagesToLinks()
+        {
+            _log.LogTrace("Offering messages to links");
+            List<TargetLink> linksToOfferTo;
+            lock (_lock)
+            {
+                linksToOfferTo = _targets.Where(x => x.LastOfferedMessageReply != DataflowMessageStatus.Postponed).ToList();
+            }
+
+            foreach (var link in linksToOfferTo)
+            {
+                DataflowMessageStatus reply;
+                do // keep feeding the link messages until queue is either empty or it tells us that it can't handle any more
```

Comment:
Might the developer not expect the block of have broadcast semantics? For example:
```csharp
var deltas = informer.ToResourceEventDeltaBlock(...);
deltas.LinkTo(new ActionBlock<...>(Log));
deltas.LinkTo(new ActionBlock<...>(BuildState));
```
I would expect `Log` and `BuildState` to process all deltas.

Reply:
From TPL perspective, there's no reason for them to expect broadcast behavior as messages have to be explicitly routed through a Broadcast block and "cloned" as access to the message is supposed to be thread-safe in Actor programming. None of the other TPL blocks exhibit implicit broadcast semantics. If they wanted to do broadcast they can always do `.Publish().AutoConnect()` prior to calling `ToResourceEventDeltaBlock`.